src/firmware/app/primitives/primitive_manager.c

```diff
@@ -192,21 +177,30 @@ void app_primitive_manager_runCurrentPrimitive(PrimitiveManager_t *manager,
 {
     app_primitive_manager_lockPrimitiveMutex(manager);
 
-    if (manager->current_primitive_functions)
+    if (manager->current_primitive)
     {
-        manager->current_primitive_functions->tick(manager->current_primitive_state,
-                                                   world);
+        manager->current_primitive->tick(manager->current_primitive_state, world);
     }
 
     app_primitive_manager_unlockPrimitiveMutex(manager);
 }
 
-unsigned int app_primitive_manager_getCurrentPrimitiveIndex(PrimitiveManager_t *manager)
+void app_primitive_manager_endCurrentPrimitive(PrimitiveManager_t *manager,
+                                               FirmwareWorld_t *world)
 {
-    return manager->current_primitive_index;
-}
+    if (manager->current_primitive)
+    {
+        manager->current_primitive->end(manager->current_primitive_state, world);
+        manager->current_primitive->destroy_state(manager->current_primitive_state);
 
-bool app_primitive_manager_primitiveIsDirect(unsigned int primitive)
-{
-    return PRIMITIVES[primitive]->direct;
+        manager->current_primitive = NULL;
+    }
+
+    FirmwareRobot_t *robot = app_firmware_world_getRobot(world);
+    Chicker_t *chicker     = app_firmware_robot_getChicker(robot);
+    Dribbler_t *dribbler   = app_firmware_robot_getDribbler(robot);
+
+    app_chicker_disableAutochip(chicker);
```

Comment:
Should we also set the `STOP_PRIMITIVE` here? If we're ending a primitive because we received a different one, we shouldn't see the behavior of the stop primitive since the new one will be set before the next tick (correct me if I'm wrong). But if for whatever reason we don't have another primitive, we don't want to continue doing anything from the previous primitive, so we should stop to be safe.

Reply:
This is a great point. Added the following function that is called here and in the `default` case of the switch above.
```
/**
 * Make the robot in the given world "safe" by disabling potentially dangerous 
 * functionality, including bringing it to a stop
 *
 * @param manager [in/out] The primitive manager controlling the robot
 * @param world [in] The world containing the robot make safe
 */
void app_primitive_manager_makeRobotSafe(PrimitiveManager_t *manager,
                                         FirmwareWorld_t *world);
=
```